src/search/FindReplace.js

```diff
@@ -494,6 +576,82 @@ define(function (require, exports, module) {
+        var tickIndex;
+        for (tickIndex = 0; tickIndex < lineMatchPatternArray.length; tickIndex++) {
+            if (lineMatchPatternArray[tickIndex]) {
+                scrollTrackPositions.push({line: Math.ceil(tickIndex * linesPerArraySlot)});
+            }
+        }
+
+        state.tickMarks = scrollTrackPositions;
+        ScrollTrackMarkers.clear();
+        ScrollTrackMarkers.addTickmarks(editor, scrollTrackPositions);
+    }, 100);
+
+    function indicateHasMatches(state, numResults) {
+        // Make the field red if it's not blank and it has no matches (which also covers invalid regexes)
+        findBar.showNoResults(!state.foundAny && findBar.getQueryInfo().query);
+
+        // Navigation buttons enabled if we have a query and more than one match
+        findBar.enableNavigation(state.foundAny && numResults > 1);
+        findBar.enableReplace(state.foundAny);
+    }
+
+    function performSearchUsingCursor(cm, state, editor) {
+        cm.operation(function () {
+            var cursor = state.updateSearchCursor(cm, state.searchStartPos);
+            state.optimizeForLargeDocuments = cursor.getDocCharacterCount() > FIND_OPTIMIZE_THRESHOLD;
+            var resultCount = cursor.scanDocumentAndStoreResultsInCursor();
+
+            // Highlight all matches if there aren't too many
+            if (resultCount <= FIND_HIGHLIGHT_MAX) {
+                toggleHighlighting(editor, true);
+
+                cursor.forEachMatch(function (fromPos, toPos) {
+                    var mark = cm.markText(fromPos, toPos, { className: "CodeMirror-searching", startStyle: "searching-first", endStyle: "searching-last" });
+                    state.marked.push(mark);
+                });
+
+            } else {
+                // with a lot of matches, we will highlight dynamically in the viewport
+                // as the user scrolls.
+                toggleHighlighting(editor, true);
+                // first, highlight what the user is currently viewing
+                highlightMatchesWithinViewport(cm, state);
+                // start the event listener to highlight on scroll
+                enableViewportHighlightingOfCurrentMatches(cm, editor, state);
+            }
+
+            addScrollTicksDebounced(editor);
+
+            // Here we only update find bar with no result. In the case of a match
+            // a findNext() call is guaranteed to be followed by this function call,
+            // and findNext() in turn calls _updateFindBarWithMatchInfo() to show the
+            // match index.
+            if (state.searchCursor.getMatchCount() === 0) {
+                findBar.showFindCount(Strings.FIND_NO_RESULTS);
+            }
+
+            state.foundAny = (state.searchCursor.getMatchCount() > 0);
+            indicateHasMatches(state, state.searchCursor.getMatchCount());
+
+        });
+    }
+
+    var performSearchUsingCursorDebounced = _.debounce(performSearchUsingCursor, 0);
```

Comment:
@dakaraphi did you have any specific rationale in mind for debounce timeouts? I thought of it as the following: given average speed of typing is ~200 characters per minute, it yields about 300ms per character. I instrumented the code to measure and I see a lots of debounce ops actually being cancelled as I type without visually noticeable impact if this timeout is increased to 300. Whereas with 0 all the searched never have a chance of being cancelled.

At any rate, I am interested to learn what was your reasoning for choosing these specific delays. Also, in other debounced calls (drawing ticks and such), did you try to "pipeline" the operations by choosing growing delays?